Store CSRF token in a cookie

matthijskooijman opened the pull request:
This is a subset of PR #316, with just the commits needed to move the CSRF token into a cookie. This was originally just a cleanup, but it has since been shown that having the CSRF token in the session can cause dataloss when visitors submit a (lengthy) comment and the session has expired since they first loaded the article page. In that case, the CSRF token will be invalid and the request rejected rather forcefully, losing the submitted comment.

Moving the CSRF token to a cookie ensures that the token survives until the browser closes, which should prevent this dataloss.

RoukePouw commented on system/core/events.php:
```diff
@@ -256,55 +256,38 @@ function createGotoPageAndExecuteCommandButton($label, $langPageView, $command,
-		if(isset($_POST['command'])) {
-			if (!isset($_POST['csrfToken']) || $_POST['csrfToken'] != getCsrfToken()) {
+	function executePostedCommand($O_O) {
+		static $processed = false;
```
Would it make sense to add this global/static var to `$O_O`?
Feels weird to have such a state variable lying around :-)

matthijskooijman replied:
Hm, I'm not so sure. This sneakily hidden state is not ideal, exposing this state globally through `$O_O` does not seem correct either: This is really an implementation detail of `executePostedCommand`. I think it would be good to just keep it like this now, and fix this in a future refactor (the way post commands are handled is expected to change anyway, once all datatypes are converted to the "new style").

RoukePouw replied:
Cool, just wanted to double check :-)

RoukePouw commented on system/core/HyphaRequest.php:
```diff
@@ -151,6 +161,10 @@ public function getPort() {
 		}
 
+		public function isSecure() {
+			return (array_key_exists('HTTPS', $_SERVER) && $_SERVER['HTTPS'] == 'on');
```
r edundant parentheses (non blocking comment :-) )  maybe triple `===` ?

matthijskooijman replied:
I like extra parenthesis to emphasize what is being returned, though thinking about it, I can't really put down in words why it would actually help :-) I guess it's a habit that's more relevant when assigning an expression to a variable, not really needed for return. Triple === is a good idea.

matthijskooijman replied:
Fixed.

RoukePouw commented on system/core/RequestContext.php:
```diff
@@ -120,6 +124,62 @@ public function getDictionary() {
+			subsequent POST requests. If there is no token
+			yet, it will be generated automatically.
+		*/
+		public function getCsrfToken() {
```
namingwise I would expect get to be just a getter (not meddling with state) the regeneration breaks that expectation. Maybe getOrGenerateCsrfToken / retrieveCsrfToken?

matthijskooijman replied:
Good point, fixed.

RoukePouw commented on system/core/RequestContext.php:
```diff
@@ -120,6 +124,62 @@ public function getDictionary() {
+		public function validCsrfToken() {
+			$received = $this->getRequest()->getPostValue('csrfToken');
+			$expected = $this->getCsrfToken();
+			return ($received == $expected);
```
redundant parentheses, maybe triple ===?

matthijskooijman replied:
Fixed.

matthijskooijman commented:
Thanks for the review. I've fixed most of your comments. I also noticed a missing `global $O_O;` that broke login (it got left behind splitting off commits into this new PR), that's also fixed now.

matthijskooijman commented:
I did a bit more testing, and will merge this now.
